File: day04/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <set>
#include <sstream>

auto read() {
    std::fstream ifs("/Users/ecem/CLionProjects/adventofcode2022/day04/day4.txt");
    std::string line;
    std::vector<std::string> result;
    while (std::getline(ifs, line)) {
        if (line.empty())
            continue;
        result.push_back(line);
    }
    return result;
}

std::pair<int64_t, int64_t> parse(const std::string &input) {
    auto pos = input.find('-');
    assert(pos != std::string::npos);
    return std::make_pair(std::stoll(input.substr(0, pos)), std::stoll(input.substr(pos + 1)));
}

void part1() {
    int64_t result = 0;
    for (const auto &line: read()) {
        auto pos = line.find(',');
        assert(pos != std::string::npos);
        auto a = parse(line.substr(0, pos));
        auto b = parse(line.substr(pos + 1));
        if ((a.first <= b.first && a.second >= b.second) || (b.first <= a.first && b.second >= a.second))
            ++result;
    }
    std::cout << "Part1: " << result << std::endl;
}


void part2() {
    int64_t result = 0;
    for (const auto &line: read()) {
        std::istringstream ss(line);
        std::string first, second;
        std::getline(ss, first, ',');
        std::getline(ss, second, ',');
        auto a = parse(first);
        auto b = parse(second);
        if (std::min(a.second, b.second) >= std::max(a.first, b.first))
            ++result;
    }
    std::cout << "Part2: " << result << std::endl;
}

int main() {

    part1();
    part2();
    return 0;
}

File: day04/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;


fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>> where P: AsRef<Path> {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "/Users/ecem/CLionProjects/adventofcode2022/day04/day4.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn parse(s: &str) -> (i64, i64) {
    let elems: Vec<&str> = s.split('-').collect();
    return (elems[0].parse::<i64>().unwrap(), elems[1].parse::<i64>().unwrap());
}

fn part1() {
    let result: i32 = read().iter().map(|s| {
        let elems: Vec::<&str> = s.split(',').collect();
        assert_eq!(elems.len(), 2);
        let a = parse(elems[0]);
        let b = parse(elems[1]);
        if (a.0 <= b.0 && a.1 >= b.1) || (b.0 <= a.0 && b.1 >= a.1) {
            return 1;
        }
        return 0;
    }).sum();

    println!("Part1: {}", result)
}

fn part2() {
    let result: i32 = read().iter().map(|s| {
        let elems: Vec::<&str> = s.split(',').collect();
        assert_eq!(elems.len(), 2);
        let a = parse(elems[0]);
        let b = parse(elems[1]);
        if a.1.min(b.1) >= a.0.max(b.0) {
            return 1;
        }
        return 0;
    }).sum();

    println!("Part2: {}", result)
}

fn main() {
    part1();
    part2();
}


